### PR TITLE
Add basic functions of Program Pass

### PR DESCRIPTION
**Pull request by sneaxiy**
### PR types
New features

### PR changes
Others

### Describe
Add basic functions for `Pass` applied to `ProgramDesc`. More codes would be added in the next PRs.

### CODE REVIEW

File: paddle/fluid/framework/ir/graph_helper.cc [hunk 1]
```diff
@@ -475,6 +477,117 @@ std::vector<ir::Node *> TopologySortGraphByDescOrder(const Graph &graph) {
...
+    } else if (n->Op()) {
+      ops->emplace_back(*n->Op());
+    }
+    // delete no OpDesc op
+  }
+}
+
+static void GraphToBlock(const Graph &graph, proto::BlockDesc *block,
+                         const SortKind *sort_kind) {
+  // Remove the unneeded variables after memory optimization.
```
Comment (zhhsplendid) [inline paddle/fluid/framework/ir/graph_helper.cc:514]: Just for you information, nothing to change.

I don't know where did we use the kGraphToProgramVarsToRemove and `var2remove` ... I thought we can delete these code, but I am not sure so I didn't do it in my past PRs.

File: paddle/fluid/framework/ir/graph.h [hunk 2]
```diff
@@ -47,6 +47,10 @@ constexpr char kStaleProgramOpDescs[] = "stale_program_op_descs";
 
 namespace ir {
 
+const char kGraphToProgramVarsToRemove[] =
```
Comment (zhhsplendid) [inline paddle/fluid/framework/ir/graph.h]: Should we put it at `graph.h`? It is only used in GraphToProgram and your PR moved the GraphToProgram to graph_helper, can we put those constant at `graph_helper.h`?

File: python/paddle/fluid/framework.py [hunk 3]
```diff
@@ -4148,6 +4164,91 @@ def __init__(self):
         # compiled program, i.e. Graph
         self._graph = None
 
+    def _find_var_ctor_kwargs(self, new_desc):
```
Comment (zhhsplendid) [inline python/paddle/fluid/framework.py]: what's the meaning of `ctor`? From the code, you are finding 'class' and 'kwargs', can we rename the method name as `_find_var_class_kwargs`?

File: paddle/fluid/framework/ir/graph_helper.cc (hunk 1 above)
Comment (sneaxiy) [reply paddle/fluid/framework/ir/graph_helper.cc:514]: Thanks for your FYI! I agree with you. Just keep it as the old logic.

File: paddle/fluid/framework/ir/graph.h (hunk 2 above)
Comment (sneaxiy) [reply paddle/fluid/framework/ir/graph.h]: Done.

File: python/paddle/fluid/framework.py (hunk 3 above)
Comment (sneaxiy) [reply python/paddle/fluid/framework.py]: Done.

**Review by zhhsplendid — APPROVED** [review]
LGTM

File: python/paddle/fluid/framework.py [hunk 4]
```diff
@@ -4148,6 +4164,91 @@ def __init__(self):
...
+    def _find_var_class_kwargs(self, new_desc):
+        old_desc = self.desc
+        all_new_vars = []
+        block_num = new_desc.num_blocks()
+        for idx in range(block_num):
+            new_block_desc = new_desc.block(idx)
+            all_new_vars.append([])
+            block_new_vars = all_new_vars[-1]
+            for new_var_desc in new_block_desc.all_vars():
+                if self.blocks[idx].has_var(new_var_desc.name()):
```
Comment (zhiqiu) [inline python/paddle/fluid/framework.py:4176]: The block number of new_desc and old_desc is always equal?

**Review by zhiqiu — APPROVED** [review]
LGTM for the framework.py

File: python/paddle/fluid/framework.py (hunk 4 above)
Comment (sneaxiy) [reply python/paddle/fluid/framework.py:4176]: For now, it is true. We can fix this if this is not in the future.

**Pull request merged**